File: src/lib/appointment.ts
import type { Prisma } from "@prisma/client";
import { releaseBookingHolds } from "./booking";
import { appointmentInclude, bookingTotalLyd, type LoadedAppointment } from "./bridey-pass";
import { writeAudit } from "./fees";

type Tx = Prisma.TransactionClient;

export class AppointmentError extends Error {
  constructor(
    message: string,
    public status = 400,
  ) {
    super(message);
    this.name = "AppointmentError";
  }
}

export type AppointmentAction = "check_in" | "start" | "complete" | "mark_paid" | "record_payment";

async function load(tx: Tx, bookingId: string, businessId: string) {
  const booking = await tx.booking.findFirst({
    where: { id: bookingId, businessId },
    include: appointmentInclude,
  });
  if (!booking) throw new AppointmentError("NOT_FOUND", 404);
  return booking;
}

async function audit(
  tx: Tx,
  args: { artistId: string; ownerArtistId: string; action: string; bookingId: string; extra?: string },
) {
  await writeAudit(tx, {
    actorType: "artist",
    actorId: args.artistId,
    action: args.action,
    artistId: args.ownerArtistId,
    reason: `booking:${args.bookingId}${args.extra ? `;${args.extra}` : ""}`,
  });
}

export async function applyAppointmentAction(
  tx: Tx,
  args: {
    bookingId: string;
    businessId: string;
    memberId: string;
    artistId: string;
    ownerArtistId: string;
    action: AppointmentAction;
    amountLyd?: number;
    depositLyd?: number;
  },
): Promise<LoadedAppointment> {
  const current = await load(tx, args.bookingId, args.businessId);
  if (current.status === "CANCELLED" && (args.action === "check_in" || args.action === "start" || args.action === "complete")) {
    throw new AppointmentError("INVALID_STATUS", 400);
  }

  if (args.action === "check_in") {
    if (["CHECKED_IN", "IN_PROGRESS", "COMPLETED"].includes(current.status)) return current;
    if (current.status !== "CONFIRMED") throw new AppointmentError("INVALID_STATUS", 400);
    const result = await tx.booking.updateMany({
      where: { id: current.id, status: "CONFIRMED" },
      data: {
        status: "CHECKED_IN",
        checkedInAt: current.checkedInAt || new Date(),
        checkedInById: current.checkedInById || args.memberId,
      },
    });
    if (result.count === 0) {
      const again = await load(tx, args.bookingId, args.businessId);
      if (["CHECKED_IN", "IN_PROGRESS", "COMPLETED"].includes(again.status)) return again;
      throw new AppointmentError("INVALID_STATUS", 400);
    }
    await audit(tx, { artistId: args.artistId, ownerArtistId: args.ownerArtistId, action: "booking.checked_in", bookingId: current.id });
    return load(tx, args.bookingId, args.businessId);
  }

  if (args.action === "start") {
    if (["IN_PROGRESS", "COMPLETED"].includes(current.status)) return current;
    if (current.status !== "CHECKED_IN") throw new AppointmentError("INVALID_STATUS", 400);
    const result = await tx.booking.updateMany({
      where: { id: current.id, status: "CHECKED_IN" },
      data: {
        status: "IN_PROGRESS",
        startedAt: current.startedAt || new Date(),
        startedById: current.startedById || args.memberId,
      },
    });
    if (result.count === 0) {
      const again = await load(tx, args.bookingId, args.businessId);
      if (["IN_PROGRESS", "COMPLETED"].includes(again.status)) return again;
      throw new AppointmentError("INVALID_STATUS", 400);
    }
    await audit(tx, { artistId: args.artistId, ownerArtistId: args.ownerArtistId, action: "booking.started", bookingId: current.id });
    return load(tx, args.bookingId, args.businessId);
  }

  if (args.action === "complete") {
    if (current.status === "COMPLETED") return current;
    if (current.status !== "IN_PROGRESS") throw new AppointmentError("INVALID_STATUS", 400);
    const result = await tx.booking.updateMany({
      where: { id: current.id, status: "IN_PROGRESS" },
      data: {
        status: "COMPLETED",
        completedAt: current.completedAt || new Date(),
        completedById: current.completedById || args.memberId,
        expiresAt: null,
      },
    });
    if (result.count === 0) {
      const again = await load(tx, args.bookingId, args.businessId);
      if (again.status === "COMPLETED") return again;
      throw new AppointmentError("INVALID_STATUS", 400);
    }
    await releaseBookingHolds(tx, current.id);
    await audit(tx, { artistId: args.artistId, ownerArtistId: args.ownerArtistId, action: "booking.completed", bookingId: current.id });
    return load(tx, args.bookingId, args.businessId);
  }

  if (args.action === "mark_paid" || args.action === "record_payment") {
    if (!["CONFIRMED", "CHECKED_IN", "IN_PROGRESS", "COMPLETED"].includes(current.status)) {
      throw new AppointmentError("INVALID_STATUS", 400);
    }
    const total = bookingTotalLyd(current);
    const paid = current.paidLyd || 0;
    if (paid >= total && total > 0) return current;
    const add =
      args.action === "mark_paid"
        ? Math.max(0, total - paid)
        : Math.max(0, Math.floor(Number(args.amountLyd) || 0));
    if (add <= 0) return current;
    const nextPaid = Math.min(total, paid + add);
    const depositLyd =
      typeof args.depositLyd === "number" && args.depositLyd >= 0
        ? Math.floor(args.depositLyd)
        : current.depositLyd || (paid === 0 && nextPaid < total ? nextPaid : current.depositLyd);
    const result = await tx.booking.updateMany({
      where: {
        id: current.id,
        paidLyd: paid,
        status: { in: ["CONFIRMED", "CHECKED_IN", "IN_PROGRESS", "COMPLETED"] },
      },
      data: {
        paidLyd: nextPaid,
        depositLyd,
        paidAt: new Date(),
        paidById: args.memberId,
      },
    });
    if (result.count === 0) {
      return load(tx, args.bookingId, args.businessId);
    }
    await audit(tx, {
      artistId: args.artistId,
      ownerArtistId: args.ownerArtistId,
      action: "booking.payment_recorded",
      bookingId: current.id,
      extra: `paid:${nextPaid};total:${total}`,
    });
    return load(tx, args.bookingId, args.businessId);
  }

  throw new AppointmentError("INVALID", 400);
}

File: backend/bridey_api/appointment.py
from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from bridey_api.booking import release_booking_holds
from bridey_api.errors import AppointmentError
from bridey_api.fees import write_audit
from bridey_api.models import Booking, BookingAssignment
from bridey_api.pass_tokens import booking_total_lyd

APPOINTMENT_LOAD = (
    selectinload(Booking.service),
    selectinload(Booking.items),
    selectinload(Booking.shift),
    selectinload(Booking.assignments).selectinload(BookingAssignment.team_member),
    selectinload(Booking.assignments).selectinload(BookingAssignment.service),
    selectinload(Booking.fee),
)


def load_appointment(session: Session, booking_id: str, business_id: str) -> Booking:
    booking = session.scalars(
        select(Booking).where(Booking.id == booking_id, Booking.business_id == business_id).options(*APPOINTMENT_LOAD)
    ).first()
    if not booking:
        raise AppointmentError("NOT_FOUND", 404)
    return booking


def apply_appointment_action(session: Session, args: dict) -> Booking:
    current = load_appointment(session, args["bookingId"], args["businessId"])
    action = args["action"]
    if current.status == "CANCELLED" and action in {"check_in", "start", "complete"}:
        raise AppointmentError("INVALID_STATUS", 400)

    if action == "check_in":
        if current.status in {"CHECKED_IN", "IN_PROGRESS", "COMPLETED"}:
            return current
        if current.status != "CONFIRMED":
            raise AppointmentError("INVALID_STATUS", 400)
        if current.status == "CONFIRMED":
            current.status = "CHECKED_IN"
            current.checked_in_at = current.checked_in_at or datetime.now(timezone.utc)
            current.checked_in_by_id = current.checked_in_by_id or args["memberId"]
            session.flush()
        write_audit(
            session,
            {
                "actorType": "artist",
                "actorId": args["artistId"],
                "action": "booking.checked_in",
                "artistId": args["ownerArtistId"],
                "reason": f"booking:{current.id}",
            },
        )
        return load_appointment(session, args["bookingId"], args["businessId"])

    if action == "start":
        if current.status in {"IN_PROGRESS", "COMPLETED"}:
            return current
        if current.status != "CHECKED_IN":
            raise AppointmentError("INVALID_STATUS", 400)
        current.status = "IN_PROGRESS"
        current.started_at = current.started_at or datetime.now(timezone.utc)
        current.started_by_id = current.started_by_id or args["memberId"]
        session.flush()
        write_audit(
            session,
            {
                "actorType": "artist",
                "actorId": args["artistId"],
                "action": "booking.started",
                "artistId": args["ownerArtistId"],
                "reason": f"booking:{current.id}",
            },
        )
        return load_appointment(session, args["bookingId"], args["businessId"])

    if action == "complete":
        if current.status == "COMPLETED":
            return current
        if current.status != "IN_PROGRESS":
            raise AppointmentError("INVALID_STATUS", 400)
        current.status = "COMPLETED"
        current.completed_at = current.completed_at or datetime.now(timezone.utc)
        current.completed_by_id = current.completed_by_id or args["memberId"]
        current.expires_at = None
        session.flush()
        release_booking_holds(session, current.id)
        write_audit(
            session,
            {
                "actorType": "artist",
                "actorId": args["artistId"],
                "action": "booking.completed",
                "artistId": args["ownerArtistId"],
                "reason": f"booking:{current.id}",
            },
        )
        return load_appointment(session, args["bookingId"], args["businessId"])

    if action in {"mark_paid", "record_payment"}:
        if current.status not in {"CONFIRMED", "CHECKED_IN", "IN_PROGRESS", "COMPLETED"}:
            raise AppointmentError("INVALID_STATUS", 400)
        total = booking_total_lyd(current)
        paid = current.paid_lyd or 0
        if paid >= total and total > 0:
            return current
        add = max(0, total - paid) if action == "mark_paid" else max(0, int(args.get("amountLyd") or 0))
        if add <= 0:
            return current
        next_paid = min(total, paid + add)
        deposit = args.get("depositLyd")
        if isinstance(deposit, (int, float)) and deposit >= 0:
            current.deposit_lyd = int(deposit)
        elif paid == 0 and next_paid < total:
            current.deposit_lyd = next_paid
        current.paid_lyd = next_paid
        current.paid_at = datetime.now(timezone.utc)
        current.paid_by_id = args["memberId"]
        session.flush()
        write_audit(
            session,
            {
                "actorType": "artist",
                "actorId": args["artistId"],
                "action": "booking.payment_recorded",
                "artistId": args["ownerArtistId"],
                "reason": f"booking:{current.id};paid:{next_paid};total:{total}",
            },
        )
        return load_appointment(session, args["bookingId"], args["businessId"])

    raise AppointmentError("INVALID", 400)
